tests/Predis/Command/Redis/Json/JSONSET_Test.php

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * @group connected
+     * @return void
+     * @requiresRedisJsonVersion >= 1.0.0
+     */
+    public function testThrowsExceptionOnUnexpectedValueGiven(): void
+    {
+        $redis = $this->getClient();
+
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessage('Subcommand argument accepts only: nx, xx values');
+
+        $redis->jsonset('key', '$', 'value', 'wrong');
+    }
+
+    public function argumentsProvider(): array
+    {
+        return [
+            'with default arguments' => [
+                ['key', 'path', 'value'],
+                ['key', 'path', 'value']
+            ],
+            'with NX subcommand' => [
+                ['key', 'path', 'value', 'nx'],
+                ['key', 'path', 'value', 'NX'],
+            ],
+            'with XX subcommand' => [
+                ['key', 'path', 'value', 'xx'],
+                ['key', 'path', 'value', 'XX'],
+            ],
+        ];
+    }
+
+    public function jsonProvider(): array
+    {
+        return [
+            'override json' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '{"key3":"value3"}',
+                '$',
+                null,
+                'OK',
+                '{"key3":"value3"}'
+            ],
+            'override certain key - without subcommands' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '"value3"',
+                '$.key2',
+                null,
+                'OK',
+                '{"key1":"value1","key2":"value3"}'
+            ],
+            'append to json - without subcommands' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '"value3"',
+                '$.key3',
+                null,
+                'OK',
+                '{"key1":"value1","key2":"value2","key3":"value3"}'
+            ],
+            'override certain key - with XX subcommand' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '"value3"',
+                '$.key2',
+                'xx',
+                'OK',
+                '{"key1":"value1","key2":"value3"}'
+            ],
+            'append to json - with NX subcommand' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '"value3"',
+                '$.key3',
+                'nx',
+                'OK',
+                '{"key1":"value1","key2":"value2","key3":"value3"}'
+            ],
+            'override failed with XX subcommand' => [
+                'key',
+                '{"key1":"value1","key2":"value2"}',
+                '"value3"',
+                '$.key3',
+                'xx',
+                null,
+                '{"key1":"value1","key2":"value2"}'
+            ],
+            'append failed with NX subcommand' => [
```

Comment:
IMHO we should do hierarchy (path based) tests as well.

Reply:
Not sure If I get you

Reply:
redisjson allows us to not just set things at the top level. We can do path based setting like JSON. SET $ <some doc> JSON.SET $..foo <some doc>

See the redis-py tests (test_json.py) as an example.

Reply:
@chayim There's a bunch of test cases with path based JSON setting:
https://github.com/predis/predis/pull/868/files/92de6634d9ae24d7a2f6d8d6c0d6c9fba02a2980#diff-c5a3fb39f6bdb4210ce15d464750651776475b2635b1ba1be15741ffecfcf91cR120

https://github.com/predis/predis/pull/868/files/92de6634d9ae24d7a2f6d8d6c0d6c9fba02a2980#diff-c5a3fb39f6bdb4210ce15d464750651776475b2635b1ba1be15741ffecfcf91cR129

https://github.com/predis/predis/pull/868/files/92de6634d9ae24d7a2f6d8d6c0d6c9fba02a2980#diff-c5a3fb39f6bdb4210ce15d464750651776475b2635b1ba1be15741ffecfcf91cR138

https://github.com/predis/predis/pull/868/files/92de6634d9ae24d7a2f6d8d6c0d6c9fba02a2980#diff-c5a3fb39f6bdb4210ce15d464750651776475b2635b1ba1be15741ffecfcf91cR147